File: 01_정렬, 맵, 셋/라이브 코딩/2751.cpp
// N개의 수가 주어졌을 때, 이를 오름차순으로 정렬하는 프로그램을 작성하시오.

#include <iostream>
#include <vector>

using namespace std;

vector<int> sorted; // 정렬 과정에서 사용한 부분적으로 정렬된 배열

// 합병 정렬(합병)
void merge(vector<int>& arr, int left, int mid, int right) {
	int pl = left, pr = mid + 1, idx = left;
	while (pl <= mid && pr <= right) { // 왼쪽 또는 오른쪽 배열의 마지막 원소에 도달할 때까지
		if (arr[pl] < arr[pr]) {
			sorted[idx++] = arr[pl++];
		}
		else {
			sorted[idx++] = arr[pr++]
		}
	}
	while (pl <= mid) {
		sorted[idx++] = arr[pl++];
	}
	while(pr <= right) {
		sorted[idx++] = arr[pr++];
	}
	for (int i = left; i <= right; i++) {
		arr[i] = sorted[i];
	}
}

// 합병 정렬(분할): O(nlog2n)
void mergeSort(vector<int>& arr, int left, int right) {
	if (left < right) { // 원소가 하나 남을 때까지 분할
		int mid = (left + right) / 2;
		mergeSort(arr, left, mid);
		mergeSort(arr, mid + 1, right);
		merge(arr, left, mid, right);
	}
}

int main() {
	int n;
	cin >> n;
	vector<int>arr(n);
	sorted.assign(n, 0);
	for (int i = 0; i < n; i++) {
		cin >> arr[i];
	}
	mergeSort(arr,0,n-1);
	for (int i = 0; i < n; i++) {
		cout << arr[i] << ' ';
	}
	return 0;
}

File: 01_정렬, 맵, 셋/라이브 코딩/2751.py
def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    """    
    Divide: 문제 분할
    arr을 arr[:mid]와 arr[mid:]로 분할
    """
    mid = len(arr) // 2  # //는 몫을 구하는 연산자

    # Conquer: 쪼개진 작은 문제 해결
    # left와 right에는 각각 정렬된 리스트가 담김
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    i, j = 0, 0  # i는 left, j는 right의 cursor
    temp = []

    # Combine: 해결된 작은 문제들을 다시 합침
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            temp.append(left[i])
            i += 1
        else:
            temp.append(right[j])
            j += 1

    # left 혹은 right에 남은 요소들 넣기
    temp += left[i:]
    temp += right[j:]

    return temp


N = int(input())
sorted_arr = []  # 정렬된 수를 담을 리스트

for _ in range(N):
    sorted_arr.append(int(input()))
# 합병 정렬
sorted_arr = merge_sort(sorted_arr)
# 정렬된 수 출력
for n in sorted_arr:
    print(n)
